torch/utils/data/_utils/pin_memory.py

```diff
@@ -53,27 +53,13 @@ def do_one_step():
         do_one_step()
 
 def pin_memory(data, device=None):
-    if isinstance(data, torch.Tensor):
-        return data.pin_memory(device)
-    elif isinstance(data, (str, bytes)):
-        return data
-    elif isinstance(data, collections.abc.Mapping):
-        try:
-            return type(data)({k: pin_memory(sample, device) for k, sample in data.items()})  # type: ignore[call-arg]
-        except TypeError:
-            # The mapping type may not support `__init__(iterable)`.
-            return {k: pin_memory(sample, device) for k, sample in data.items()}
-    elif isinstance(data, tuple) and hasattr(data, '_fields'):  # namedtuple
-        return type(data)(*(pin_memory(sample, device) for sample in data))
-    elif isinstance(data, tuple):
-        return [pin_memory(sample, device) for sample in data]  # Backwards compatibility.
-    elif isinstance(data, collections.abc.Sequence):
-        try:
-            return type(data)([pin_memory(sample, device) for sample in data])  # type: ignore[call-arg]
-        except TypeError:
-            # The sequence type may not support `__init__(iterable)` (e.g., `range`).
-            return [pin_memory(sample, device) for sample in data]
-    elif hasattr(data, "pin_memory"):
-        return data.pin_memory()
-    else:
-        return data
+    def pin_mem(tensor):
+        if isinstance(tensor, torch.Tensor):
+            return tensor.pin_memory(device)
+        elif isinstance(tensor, (str, bytes)):
+            return tensor
+        elif hasattr(tensor, "pin_memory"):
+            return tensor.pin_memory()
+        else:
+            return tensor
+    return tree_map(pin_mem, data)
```

Comment:
I copy pasted the logic removing the containers that are cast by pytree.
To me we could just have the tensor instance check as well as the hasattr (which can introduce overhead) and then fall back on returning the tensor):
```suggestion
        if isinstance(tensor, torch.Tensor):
            return tensor.pin_memory(device)
        elif hasattr(tensor, "pin_memory"):
            return tensor.pin_memory()
        else:
            return tensor
```